File: Ahorn/entities/fancyfallingblock.jl
module FancyTileEntitiesFancyFallingBlock

using ..Ahorn, Maple
using Ahorn.AhornTileEntity

@mapdef Entity "FancyTileEntities/FancyFallingBlock" FancyFallingBlock(x::Int, y::Int,
	width::Int=Maple.defaultBlockWidth, height::Int=Maple.defaultBlockHeight,
	climbFall::Bool=true, behind::Bool=false, tileData::String="", tileDataHighlight::String="",
	manualTrigger::Bool=false,
)

const placements = Ahorn.PlacementDict(
	 "Falling Block (FancyTileEntities)" => Ahorn.EntityPlacement(
		  FancyFallingBlock,
		  "rectangle",
		Dict{String, Any}(
			"finalBoss" => false
		)
    ),
    "Badeline Boss Falling Block (FancyTileEntities)" => Ahorn.EntityPlacement(
        FancyFallingBlock,
        "rectangle",
		  Dict{String, Any}(
  			"finalBoss" => true
  		)
    ),
)

Ahorn.minimumSize(entity::FancyFallingBlock) = 8, 8
Ahorn.resizable(entity::FancyFallingBlock) = true, true

Ahorn.selection(entity::FancyFallingBlock) = Ahorn.getEntityRectangle(entity)

function Ahorn.propertyOptions(entity::FancyFallingBlock, ignores::Array{String, 1}=String[])
	push!(ignores, "finalBoss")
	if Bool(get(entity.data, "finalBoss", false))
		push!(ignores, "climbFall", "behind")
	else
		push!(ignores, "tileDataHighlight")
	end
	Ahorn.AhornTileEntity.tileEntityConfigOptions(entity, ignores)
end

Ahorn.renderAbs(ctx::Ahorn.Cairo.CairoContext, entity::FancyFallingBlock, room::Maple.Room) = renderTileEntity(ctx, entity, room)

end

File: Loenn/entities/fancy_falling_block.lua
local mods = require("mods")
local fancyTileEntitieshelper = mods.requireFromPlugin("libraries.fancy_tile_entities_helper")

local fallingBlock = {}

fallingBlock.name = "FancyTileEntities/FancyFallingBlock"
fallingBlock.placements = {
    {
        name = "falling_block",
        data = {
            tileData = "",
            climbFall = true,
            behind = false,
            manualTrigger = false,
            finalBoss = false,
            width = 8,
            height = 8
        }
    },
    {
        name = "boss_falling_block",
        data = {
            tileData = "",
            tileDataHighlight = "",
            behind = false,
            finalBoss = true,
            width = 8,
            height = 8
        }
    }
}


fallingBlock.fieldInformation = {
    tileData = {
        fieldType = "FancyTileEntities.buttonStringField"
    },
    tileDataHighlight = {
        fieldType = "FancyTileEntities.buttonStringField"
    }
}

fallingBlock.sprite = fancyTileEntitieshelper.getEntitySpriteFunction("blendEdges", "tilesFg", {1, 1, 1, 1})

fallingBlock.ignoredFields = {
    "finalBoss"
}

function fallingBlock.depth(room, entity)
    return entity.behind and 5000 or 0
end

return fallingBlock